ENH: sparse.linalg: `LinearOperator` dtype determination in `_interface.py`

lobpcg opened the pull request:
#### Reference issue
Closes gh-19207

#### What does this implement/fix?
When determining the `dtype`:
Make `dtype` the smallest possible.

#### BOGO
Adds an ability for ARPACK to work with integer type matrices, previously resulting in a Not Implemented type error.

rkern commented:
Reducing the dtype of the test vector is sensible; otherwise, they essentially all come out as `float64`. But the other two are unnecessary given a correctly-implemented `LinearOperator`.

lobpcg commented:
Using `matmat` substituting `matvec` for dtype determination fails above; see https://github.com/scipy/scipy/issues/19207#issuecomment-1712842372
So revert the change.

lobpcg commented:
I have also tried to include `rmatvec` or `rmatmat` in dtype determination in
```
            #if self.rmatvec is not None:
            #    v = np.zeros((self.shape[0],), dtype=np.int8)
            #    rdtype = np.asarray(self.rmatvec(v)).dtype
            #    self.dtype = np.promote_types(self.dtype, rdtype)
```
but they are not None if not present, as I have anticipated and hoped for, instead giving the `Not Implemented` error... So I give up.

lobpcg commented:
Ready for review please

rkern commented on scipy/optimize/_trustregion_constr/canonical_constraint.py:
```diff
@@ -133,7 +133,7 @@ def hess(x, v_eq, v_ineq):
 
             def matvec(p):
-                result = np.zeros_like(p)
+                result = np.zeros_like(p, dtype=float)
```
What if the input is `float32`?

lobpcg replied:
In the test, even with `p.dtype=int8`, the `result` is already `float64` giving the error in `+=`

rkern replied:
Which test? How is it even getting an `int8` array to begin with? Since the `dtype` is explicitly set in the constructor a few lines down, the dtype-guessing code shouldn't activate.

But since it *is* explicitly instantiated with `dtype=float`, this change is a good one. I'm just puzzled as to how you encountered it.

lobpcg replied:
the parameter p is an input to the LinearOperator and could in theory be of any type. My change made it int8 and resulted in a type mismatch error adding int8 and float64 with += failing a test. I have not paid attention to anything else around. You could find the error in the log.

rkern replied:
Ah, it's not this `LinearOperator` that's failing to determine the `dtype`, but the one inside of `LagrangianHessian` which is composed from these.

I'm satisfied that this is a good change now that I know where the failure arose.

rkern commented on scipy/sparse/linalg/_eigen/arpack/arpack.py:
```diff
@@ -319,7 +319,12 @@ def __init__(self, n, k, tp, mode=1, sigma=None,
+            if np.can_cast(tp, 'f'):
+                tp = 'f'
+            elif np.can_cast(tp, 'd'):
+                tp = 'd'
```
I would probably default to `float64` in the case of a non-floating-point input since that was the de-facto result of the old `dtype`-determination. Defaulting to `float32` caused a visible change in the precision of the outputs in the `svds()` docstring, which you had to adjust the `atol` for. That's undesirable.

lobpcg replied:
The  `svds()` docstring example is matrix-free, but both `matvec` and `rmatvec` there would correspond to multiplying by an integer-valued matrix resulting in the integer dtype of `aslinearoperator` consistently with the new behavior of the custom `LinearOperator`. 

The `svds()` docstring example calls ARPACK which before would give an error for integer types.  Now it works with integer types, calling the `float32` libraries. Maybe I should add the corresponding comments to the docstring example (which coincidently I am an author of...) explaining this rather than defaulting ARPACK to `float64` libraries for integer types while at the same time using `float32` libraries for `float32` types that is not quite logical? 

I do not think we need to be too much concerned here with backward compatibility, since such examples of `matvec` and `rmatvec` as used in this  `svds()` docstring example are very uncommon.

rkern replied:
When a function is implemented for both `float32` and `float64` and is given an integer input, we almost always use the `float64` codepath by default. `svds()` should do the same.

lobpcg replied:
Looks like I have already resolved this? @rkern

rkern commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -178,8 +178,15 @@ def _init_dtype(self):
-            v = np.zeros(self.shape[-1])
+            v = np.zeros(self.shape[-1], dtype=np.int8)
             self.dtype = np.asarray(self.matvec(v)).dtype
+            # the code below fails
```
Delete commented-out code.

lobpcg replied:
I was hoping that someone could suggest an easy fix... It would be beneficial to use `matmat` instead of `matvec` and to also involve `rmatmat`

rkern replied:
Generally, I think we should assume that, when a combination of these methods are provided, that they are consistent with one another. Or people can bypass this and just provide the `dtype` explicitly.

lobpcg replied:
Using `matmat` would instantly fix the problem reported in https://github.com/scipy/scipy/issues/19212 ...

If `matmat` is used correctly in the commented out code, but results in errors, there are bugs in the codes giving the errors, see https://github.com/scipy/scipy/issues/19207#issuecomment-1712842372 ?

rkern replied:
No, it would not have solved the problem in #19212. The problem there was that you had an invalid `LinearOperator`. Even if you got the `dtype` determination to work by avoiding calling `matvec()` *there*, in that one place, that `matvec()` would still be broken elsewhere. The only way to fix #19212 is to fix `_CustomLinearOperator` to accept at least one of `matvec` or `matmat`, as I explained. Then you can implement just `matmat` and get a fully-functional `LinearOperator`.

The errors you report in that comment are exactly the same kind of error: their `matvec` implementations are invalid so calling `matmat()` (from *anywhere*, not just the `dtype` determination) are broken as well.

rkern commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -178,8 +178,15 @@ def _init_dtype(self):
         """
         if self.dtype is None:
-            v = np.zeros(self.shape[-1])
+            v = np.zeros(self.shape[-1], dtype=np.int8)
```
Leave a comment here explaining why we use `int8`, otherwise, it's likely a later developer will change it.

lobpcg replied:
added to the function docstring above

rkern commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -175,10 +175,18 @@ def __init__(self, dtype, shape):
 
     def _init_dtype(self):
-        """Called from subclasses at the end of the __init__ routine.
+        """Determine the dtype from `matvec` by execting it on `int8`.
```
```suggestion
        """Determine the dtype from `matvec` by executing it on `int8` test vector.
```

rkern commented:
I don't think so. `LinearOperator` objects will always have a `.dtype` attribute set (unless if one writes a particularly broken subclass).

rkern commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -175,10 +175,18 @@ def __init__(self, dtype, shape):
 
     def _init_dtype(self):
-        """Called from subclasses at the end of the __init__ routine.
+        """Determine the dtype by execting `matvec` on an `int8` test vector.
```
```suggestion
        """Determine the dtype by executing `matvec` on an `int8` test vector.
```

lobpcg commented:
> I don't think so. `LinearOperator` objects will always have a `.dtype` attribute set (unless if one writes a particularly broken subclass).

Maybe remove it then as never being executed? It is inconsistent with my changes following the original framework.

rkern commented:
Oh, that's the code that already exists? Yes, it should be removed and replaced with just `aslinearoperator()` wherever it is called. It's a very old function that predates the dtype inference now incorporated directly into `LinearOperator`. It's harmless, as it will never take that branch, but also useless.

lobpcg commented:
> Oh, that's the code that already exists? Yes, it should be removed and replaced with just `aslinearoperator()` wherever it is called. It's a very old function that predates the dtype inference now incorporated directly into `LinearOperator`. It's harmless, as it will never take that branch, but also useless.

In scipy/sparse/linalg/_eigen/arpack/arpack.py the following
function
```def _aslinearoperator_with_dtype(m):```
is being called 31 times... So maybe should be changed in a separate PR after this one is merged.

lobpcg commented:
> Oh, that's the code that already exists? Yes, it should be removed and replaced with just `aslinearoperator()` wherever it is called. It's a very old function that predates the dtype inference now incorporated directly into `LinearOperator`. It's harmless, as it will never take that branch, but also useless.

I have first thought it would be better for a separate PR but waiting for the final review have just done it.

lobpcg commented:
Ready to merge?

lobpcg commented:
This PR is 4 months old now. Could someone review it and consider merging?

lobpcg commented:
I already forgot about this PR... Could someone please review it, otherwise it might just slip through the cracks.

lobpcg commented:
Trying to jumpstart it back to life after months of deep sleep

lobpcg commented:
@j-bowhay could you please have another look? It appears stalled but I would love to clear all my old PRs ready to merge before starting new ones.

j-bowhay commented on scipy/sparse/linalg/_eigen/arpack/arpack.py:
```diff
@@ -321,7 +321,11 @@ def __init__(self, n, k, tp, mode=1, sigma=None,
 
         if tp not in 'fdFD':
-            raise ValueError("matrix type must be 'f', 'd', 'F', or 'D'")
+            # Use `float64` libraties from integer dtypes.
```
Typo "libraties"? Couldn't quite work out what this was meant be

lobpcg replied:
Yes, a typo, "libraries".  This module is not a solver per se but an API to one of the precompiled libraires chosen depending on the types of the input parameters. Here, for the integer type of the input matrix, earlier unsupported, I call the same library as for the 'float64' matrix, as requested in https://github.com/scipy/scipy/pull/19209#discussion_r1320852204 , to my recollection

j-bowhay replied:
```suggestion
            # Use `float64` libraries from integer dtypes.
```

j-bowhay reviewed (commented):
Looks in good shape modulo one possible typo

j-bowhay reviewed (changes requested):
Could you add a test for the type determination, eg. for this new integer functionality

lobpcg commented:
> Could you add a test for the type determination, eg. for this new integer functionality

There is already a new test in this PR as I can see. Could you please elaborate your request?

j-bowhay commented:
> > Could you add a test for the type determination, eg. for this new integer functionality
> 
> There is already a new test in this PR as I can see. Could you please elaborate your request?

The new tests are for the changes to `svds` as far as I could see. Could you add a test like the example in your original issue
```    >>> import numpy as np
    >>> from scipy.sparse.linalg import LinearOperator
    >>> def mv(v):
    ...     return np.array([2*v[0], 3*v[1]])
    ...
    >>> A = LinearOperator((2,2), matvec=mv)
 ```

lobpcg commented:
@j-bowhay et al.
While waiting so long to be reviewed, the code has become outdated due to the introduction of new rules https://numpy.org/neps/nep-0050-scalar-promotion.html#proposed-weak-promotion  in NumPy 2. Too large for int8 generic NumPy and python integers used to perform the matvec on an int8 vector now result in errors 'Python integer 200 out of bounds for int8'. In the current revision of the PR, the errors come from doc/source/tutorial/optimize.rst where such integers are used in examples.  

The core function 

```
    def _init_dtype(self):
        """Determine the dtype by executing `matvec` on an `int8` test vector.

        In `np.promote_types` hierarchy, the type `int8` is the smallest,
        so we call `matvec` on `int8` and use the promoted dtype of the output
        to set the default `dtype` of the `LinearOperator`.
        We assume that `matmat`, `rmatvec`, and `rmatmat` would result in
        the same dtype of the output given an `int8` input as `matvec`.

        Called from subclasses at the end of the __init__ routine.
        """
        if self.dtype is None:
            v = np.zeros(self.shape[-1], dtype=np.int8)
            self.dtype = np.asarray(self.matvec(v)).dtype
```

in scipy/sparse/linalg/_interface.py thus needs some more sophisticated logic, I guess. Suggestions are very welcome to fix this new issue, please.

lobpcg commented:
@j-bowhay could you please have another look? I have added the missing pytests per your request and had to massage the logic of the main function to deal with the overflow of the Python int being converted into one of numpy ints in numpy 2.

My pytests all pass know. The only failed run appears unrelated to this PR.

lobpcg commented:
While idling, I forgot how to run properly pytest on my local setup so had to burn a bit of CI resources here, sorry.

j-bowhay commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -175,11 +175,29 @@ def __init__(self, dtype, shape):
+            v = np.zeros(self.shape[-1], dtype=np.int8)
+            try:
+                matvec_v = np.asarray(self.matvec(v))
+            except OverflowError:
+                # Python large `int` promoted to `np.int64`or `np.int32`
+                try:
+                    test_int64 = np.int64(1)  # noqa: F841
+                    self.dtype = np.int64
+                except AttributeError:
+                    self.dtype = np.int32
```
I don't know if this will work because a python int can still overflow a long int?
```
In [2]: 10**50*np.array(1, dtype=np.int64)
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
Cell In[2], line 1
----> 1 10**50*np.array(1, dtype=np.int64)

OverflowError: Python int too large to convert to C long
```

lobpcg replied:
Thanks, great point! 

Maybe I should just give up and turn Python int outside of int8 into a float64? It would also make the code backward compatible. What do you think?

lobpcg replied:
testing turn Python int outside of int8 into a float64

lobpcg replied:
@j-bowhay casting Python int outside of int8 into a float64 is not always working, so I am back to my original code. It does not cast all the cases accurately, as you point out, but it is still an improvement over the main that is always `np.float64`. I have added a comment to the docs to underscore the possible inaccuracy.

j-bowhay commented:
@lobpcg, thanks for pushing this forward! I am away travelling for the next four months without much computer access, so unfortunately, it might be a while before I can look at this again. Maybe @rkern would kindly be able to take another look.

rkern reviewed (approved):
One spelling correction and a question about an area where the code can be simplified.

With the documentation recommending explicitly specifying the dtype, I'm happy enough with the remaining edge cases.

lobpcg commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -175,11 +183,29 @@ def __init__(self, dtype, shape):
+                try:
+                    test_int64 = np.int64(1)  # noqa: F841
+                    self.dtype = np.int64
+                except AttributeError:
```
without it, there were pytest errors coming above from (some or all?) 32-bit systems

rkern replied:
Since you currently use `np.int64` explicitly in the tests, I don't think it can be the case that the `AttributeError` you saw was coming from the expression `np.int64`.

Again, if you just want the default integer type for the platform, just use `np.dtype(int)`.

lobpcg replied:
The pytest currently runs over
`INT_DTYPES = (np.int8, np.int16, np.int32, np.int64)`

@rkern I am unsure how to implement your suggestion, unfortunately. Could you please propose the specific change that I would just accept and thus check if it passes?

lobpcg commented on scipy/sparse/linalg/tests/test_interface.py:
```diff
@@ -438,6 +439,32 @@ def matvec(v):
+REAL_DTYPES = (np.float32, np.float64, np.longdouble)
+COMPLEX_DTYPES = (np.complex64, np.complex128, np.clongdouble)
+INEXECTDTYPES = REAL_DTYPES + COMPLEX_DTYPES
+ALLDTYPES = INT_DTYPES + INEXECTDTYPES
```
@rkern I am quite sure that I have copy/pasted it from existing scipy pytests so the same type should likely be present in the main. 

Should I search for it and fix it in this PR if found?

lobpcg commented:
@rkern I do not know why one test has started failing https://github.com/scipy/scipy/actions/runs/11410097437/job/31751716717?pr=19209
after fixing the typo in the unrelated test... The fails are in tests of "asarray" which is concerning but still should be unrelated to this PR?

lobpcg commented:
@rkern everything fixed

rkern commented on scipy/sparse/linalg/_interface.py:
```diff
@@ -175,11 +183,29 @@ def __init__(self, dtype, shape):
+            try:
+                matvec_v = np.asarray(self.matvec(v))
+            except OverflowError:
+                # Python large `int` promoted to `np.int64`or `np.int32`
+                try:
+                    test_int64 = np.int64(1)  # noqa: F841
+                    self.dtype = np.int64
+                except AttributeError:
+                    self.dtype = np.int32
```
```suggestion
                # Python large `int` promoted to `np.int64`or `np.int32`
                self.dtype = np.dtype(int)
```

lobpcg replied:
Thanks very much! This seems to work, with the failure in the benchmark that appears unrelated.

rkern commented:
Agreed. In it goes! Thank you for your patience and endurance.
